Enumerate fix for c++17, c++14, c++11 usage

permcody opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

If somebody currently compiles MOOSE with -std=c++17, they will
get errors from use of enumerate. This is because the utility changes
the return type of enumerate conditionally based on the compiler standard
in use, which is a really poor idea. This PR makes the interface a little
less nice but consistently returns a tuple of the index and the aribtrary
type. The c++17 structured binding use is unchanged and very nice looking.
The c++14 and below usage however requires us to dig out the types with
the std::get<>() calls, which is not quite as nice. At least we'll
be ready to move forward on syntax when the time comes and now we can
compile under different standards without issue.

permcody commented:
Tag @milljm - This will fix your Conda issue on the Linux box.

moosebuild commented:
Job [App tests PETSc alt](https://civet.inl.gov/job/439313/) on f1e8426 : invalidated by @permcody

mobuild58 seems to be dead

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/439314/) on f1e8426 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14470/site)

This comment will be updated on new commits.

permcody committed:
Fixup usage of enumerate utility

Currently, if somebody compiles MOOSE with -std=c++17, they will
get errors from use of Moose::enumerate. This is because the utility changes
the return type of enumerate conditionally based on the compiler standard
in use, which is a really poor idea. This PR returns the C++11 compatible
shim class, which won't quite work with c++17. We'll have to change the
calling code anyway in the future.

refs #13126

milljm reviewed (approved):
👍
